tools/userland-zone

```diff
@@ -0,0 +1,261 @@
+
+    try:
+        logger.debug('Executing "%s"', " ".join([cmd] + args))
+        result = subprocess.run([cmd] + args, **kwargs)
+    except subprocess.CalledProcessError as e:
+        logger.error(
+            'Command "%s" exited with %s', " ".join([cmd] + args), e.returncode
+        )
+    except OSError as e:
+        if e.errno == errno.ENOENT:
+            raise ValueError("{} cannot be found".format(cmd))
+
+    if result and result.stdout:
+        ret = result.stdout.decode()
+        ret = ret.rstrip()
+
+    return ret
+
+
+def parse_arguments():
+    def dir_path(path):
+        if os.path.exists(path) and os.path.isdir(path):
+            return path
+        raise argparse.ArgumentTypeError("{} is not a valid path".format(path))
+
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers(title="subcommands", dest="subcommand")
+
+    # create-template
+    ct_parser = subparsers.add_parser(
+        "create-template", help="Create template zone for userland builds"
+    )
+    ct_parser.add_argument(
+        "-p",
+        help="Zone path",
+        default="/zones/{}".format(TEMPLATE_ZONE_NAME),
+        dest="zone_path",
+    )
+
+    # update-template
+    _ = subparsers.add_parser(
+        "update-template", help="Update template zone for userland builds"
+    )
+
+    # destroy-template
+    _ = subparsers.add_parser(
+        "destroy-template", help="Destroy template zone for userland builds"
+    )
+
+    # spawn-zone
+    sz_parser = subparsers.add_parser(
+        "spawn-zone", help="Spawn build zone for userland builds"
+    )
+    sz_parser.add_argument(
+        "--id", required=True, help="Zone identifier that identifies build zone"
+    )
+    sz_parser.add_argument(
+        "--archive-dir",
+        default="/ws/archives",
+        type=dir_path,
+        help="Path to userland achrives",
+    )
+
+    # destroy-zone
+    dz_parser = subparsers.add_parser("destroy-zone", help="Destroy build zone")
+    dz_parser.add_argument(
+        "--id", required=True, help="Zone identifier that identifies build zone"
+    )
+
+    args = parser.parse_args()
+
+    if not args.subcommand:
+        parser.print_help()
+        exit(1)
+
+    return args
+
+
+def create_template(zone_path, zone_name=TEMPLATE_ZONE_NAME):
+    zone = Zone(path=zone_path, name=zone_name)
+    zone.create()
+    zone.install()
+    zone.configure()
+
+
+def destroy_zone(zone_name=TEMPLATE_ZONE_NAME):
+    zone = Zone(path=os.path.join(ZONES_ROOT, zone_name), name=zone_name)
+
+    if zone.is_running:
+        zone.halt()
+    zone.uninstall()
+    zone.delete()
+
+
+def spawn_zone(id, archive_dir="/ws/archives"):
+    name = "{}-{}".format(BUILD_ZONE_NAME, id)
+    zone = Zone(name=name, path=os.path.join(ZONES_ROOT, name))
+
+    template_zone = Zone(name=TEMPLATE_ZONE_NAME)
+
+    extra_args = [
+        "add fs",
+        "set dir = /ws/archives",
+        "set special = {}".format(archive_dir),
```

Comment:
I prefer to have the dir as a constant like BUILD_ZONE_NAME and have the configuration items in one place.